File: solutions/python3/0068.py
class Solution:
    def fullJustify(self, words: List[str], maxWidth: int) -> List[str]:
        ret = []
        curr = []
        numOfLetters = 0

        for word in words:
            if numOfLetters + len(word) + len(curr) > maxWidth:
                for i in range(maxWidth - numOfLetters):
                    curr[i % (len(curr) - 1 or 1)] += ' '
                ret.append(''.join(curr))
                curr = []
                numOfLetters = 0
            curr.append(word)
            numOfLetters += len(word)

        return ret + [' '.join(curr).ljust(maxWidth)]

File: solutions/cpp/0068.cpp
class Solution {
public:
    vector<string> fullJustify(vector<string>& words, int maxWidth) {
        vector<string> ret;
        vector<string> curr;
        int numOfLetters = 0;

        for (const auto& word : words) {
            if (numOfLetters + (int)curr.size() + (int)word.size() > maxWidth) {
                for (int i = 0; i < maxWidth - numOfLetters; i++) {
                    curr.size() - 1 == 0
                        ? curr[0].append(" ")
                        : curr[i % (curr.size() - 1)].append(" ");
                }
                ret.push_back(join(curr, ""));
                curr.clear();
                numOfLetters = 0;
            }
            curr.push_back(word);
            numOfLetters += word.size();
        }
        ret.push_back(ljust(join(curr, " "), maxWidth));

        return ret;
    }

private:
    string join(const vector<string>& v, const string c) {
        string s;
        for (auto p = v.begin(); p != v.end(); p++) {
            s += *p;
            if (p != v.end() - 1) s += c;
        }
        return s;
    }

    string ljust(string s, int width) {
        for (int i = 0; i < s.size() - width; i++) s += " ";
        return s;
    }
};